public/app/core/utils/UserProvider.tsx

```diff
@@ -15,36 +20,75 @@ export interface ChangePasswordFields {
   confirmNew: string;
 }
 
+export interface ProfileUpdateFields {
+  name: string;
+  email: string;
+  login: string;
+}
+
 export interface Props {
-  children: (api: UserAPI, states: LoadingStates) => JSX.Element;
+  userId?: number; // passed, will load user on mount
+  children: (user: User, api: UserAPI, states: LoadingStates) => JSX.Element;
 }
 
 export interface State {
+  user?: User;
   loadingStates: LoadingStates;
 }
 
 export class UserProvider extends PureComponent<Props, State> {
   state: State = {
     loadingStates: {
       changePassword: false,
+      loadUser: true,
+      updateUserProfile: false,
     },
   };
 
+  componentDidMount() {
+    if (this.props.userId) {
+      this.loadUser();
+    }
+  }
+
   changePassword = async (payload: ChangePasswordFields) => {
     this.setState({ loadingStates: { ...this.state.loadingStates, changePassword: true } });
     await getBackendSrv().put('/api/user/password', payload);
     this.setState({ loadingStates: { ...this.state.loadingStates, changePassword: false } });
   };
 
+  loadUser = async () => {
+    this.setState({
+      loadingStates: { ...this.state.loadingStates, loadUser: true },
+    });
+    const user = await getBackendSrv().get('/api/user');
+    this.setState({ user, loadingStates: { ...this.state.loadingStates, loadUser: Object.keys(user).length === 0 } });
+  };
+
+  updateUserProfile = async (payload: ProfileUpdateFields) => {
+    this.setState({ loadingStates: { ...this.state.loadingStates, updateUserProfile: true } });
+    await getBackendSrv()
+      .put('/api/user', payload)
+      .then(() => {
+        this.loadUser();
+      })
+      .catch(e => console.log(e))
+      .finally(() => {
+        this.setState({ loadingStates: { ...this.state.loadingStates, updateUserProfile: false } });
+      });
+  };
+
   render() {
     const { children } = this.props;
-    const { loadingStates } = this.state;
+    const { loadingStates, user } = this.state;
 
     const api = {
       changePassword: this.changePassword,
+      loadUser: this.loadUser,
+      updateUserProfile: this.updateUserProfile,
     };
 
-    return <>{children(api, loadingStates)}</>;
+    return <>{children(user, api, loadingStates)}</>;
```

Comment:
It's interesting that TS is not complaining about this line. `user` is an optional field in the `State` interface, but is required in the render prop. Can you investigate? I think `user` should be optional in the render prop as well.

Reply:
I'm not sure I know what you mean.

Reply:
this is your state definition:
```tsx
export interface State {
  user?: User;
  loadingStates: LoadingStates;	
}
```

and here is the render prop definition:

```tsx
children: (user: User, api: UserAPI, states: LoadingStates) => JSX.Element;
```

As you can see, the render prop `user` has more strict typing than `user` in state. Even though, Typescript is not complaining when you do follwing in render:

```tsx
const { loadingStates, user } = this.state;
// user can be undefined here....
return children(user, ....)
```.

Since user CAN be undefined, your render prop definition should take that into account.